### ISSUE TITLE
[RFC] Introduce `accelerator_strategy` flag to Trainer

### ISSUE THREAD

**Issue by kaushikb11** [PyTorchLightning/pytorch-lightning#9053]
### Motivation

The motivation is to have a separate strategy flag to support passing training type aliases (ddp, ddp_spawn, etc) and custom TrainingTypePlugin objects.

```python
Trainer(accelerator_strategy="ddp", accelerator="gpu", devices=4)
```

At the moment, there’s a single flag accelerator tied for Accelerators as well as Training Type plugins. We wish to have them decoupled!

```python
trainer = Trainer(accelerator=GPUaccelerator(..))
trainer = Trainer(accelerator='ddp_spawn')
```

Related PR: #8597 

If you **agree** with this change, react with 🎉, if not then 🙅🏽 with comments.

#### Alternate flags to set Training Types

- `accelerator`
    - type: `Optional[Union[str, Accelerator]]` = None
    - Supports training types and Accelerator Objects
- `distributed_backend`
    - type: `Optional[str]` = None
    - Deprecated, should use `accelerator` instead
- `plugins`
    - type: `Optional[Union[List[Union[Plugin, ClusterEnvironment, str]], Plugin, ClusterEnvironment, str]]` = None
    - Supports custom lightning plugins & environment

#### What's the difference between passing training type to accelerator, distributed_backend, or plugins?

- `accelerator` and `distributed_backend` only support `DistributedType`, whereas `plugins` support Custom Training Types.

#### Exceptions:

* `Trainer(distributed_backend="ddp_cpu", accelerator_strategy="ddp_spawn")`
* `Trainer(accelerator="ddp", accelerator_strategy="ddp_spawn")`
* `Trainer(plugins="ddp_find_unused_parameters_false", accelerator_strategy="ddp_spawn")`

#### Deprecations: (Deprecated in v1.5 & will be removed in v1.6)

* Passing training type to `accelerator` flag
* Passing training type to `plugins` flag

### Alternatives

* Only deprecate passing the TrainingTypePlugin into the `plugins` argument not the `accelerator` argument.
* Use simpler `strategy` argument instead of `accelerator_strategy`.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

**Comment by ananthsub** [comment]
FYI @four4fish

**Comment by yifuwang** [comment]
Hi @kaushikb11, thanks for the proposal! It makes perfect sense.

I wonder if we can use terms other than `accelerator`/`accelerator_strategy`, so that the concepts are better prepared for model parallel use cases.

`accelerator` implies that it's the component for accelerating the training (i.e. making the trainer run faster). It made more sense when `TrainingTypePlugin`-s were mostly about `DistributedDataParallel`. However, it started to feel off with the introduction of `ddp_fully_sharded`. With model parallel use cases, often the component's responsibility is not to make the training faster, but to overcome the resource limitation (i.e. memory) of a single device/host.

If possible, I think `parallelizer`/`parallelization_strategy` would be the better terms. IMHO they feel very natural since they are directly borrowed from "data parallelism" and "model parallelism", and they cover both acceleration and scaling.

**Comment by ananthsub** [comment]
@kaushikb11 - I really like the proposal. Regarding `accelerator`, what happens to device-related related arguments on the Trainer constructor?
- `num_processes`
- `gpus`
- `auto_select_gpus`
- `tpu_cores`
- `ipus`

Will these be deprecated? I would strongly prefer we have 1 way to specify different devices or distributed training, and the `device_type`/`strategy`/`device_ids` distinction is highly extensible compared to having separate flags for every new device. For instance, as a user I would find it very confusing if I can specify both `accelerator="gpu"` and `gpus` and `tpu_cores` all at the same time. 

@yifuwang - what about `accelerator` as intended for hardware acceleration, like GPU or TPUs?  In this light,' `parallelizer="gpu"` doesn't sound fully right either  as we could still be doing single-device training

**Comment by kaushikb11** [comment]
@ananthsub 

> what happens to device-related related arguments on the Trainer constructor?

I think it's a very strong user-facing API change for the Users to accept, unfortunately. Also, it's more intuitive for the user to set `gpus` or `ipus` with one flag rather than remembering two.

But the `Trainer(accelerator="x", devices="x")` combo really shines at two instances:

* When the User does `Trainer(gpus=8)` and wishes to switch to TPU, they would need to make two changes. One to change the value of `gpus` and one to change tpu_cores. With the new API, you only had to change `accelerator` from `gpu` to `tpu`.

* With `Trainer(accelerator="auto", devices=4)`, auto-selecting of an accelerator is great functionality to have with precedence (TPUs > IPUs > GPUs > CPUs). We could also support `devices=auto`, seems to be a good enhancement as well.

**Comment by kaushikb11** [comment]
@yifuwang 

> If possible, I think parallelizer/parallelization_strategy would be the better term. IMHO they feel very natural since they are directly borrowed from "data parallelism" and "model parallelism", and they cover both acceleration and scaling.

Initially, we had considered `distributed_type`/`distributed_strategy` naming for the argument. But it wouldn't make sense if `SingleDevice` plugins are passed.

**Comment by ananthsub** [comment]
> I think it's a very strong user-facing API change for the Users to accept, unfortunately. Also, it's more intuitive for the user to set gpus or ipus with one flag rather than remembering two.

This is confusing to me: with the new arguments, there are now so many ways to configure training. As a user I'd have to learn the precedence of accelerator vs setting `gpus` directly - Do I set both? Only one? If only one, which one? What happens if I set both `gpus=X` and `tpu_cores=Y` ? 

This optionality will also reflect in more framework complexity. You know best how complex the accelerator connector class has become :/ . it's one more thing to remember: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md#main-core-value-one-less-thing-to-remember

**Comment by yifuwang** [comment]
> @yifuwang - what about accelerator as intended for hardware acceleration, like GPU or TPUs? In this light,' parallelizer="gpu" doesn't sound fully right either as we could still be doing single-device training

@ananthsub, @kaushikb11 - sorry for the confusion. I wasn't suggesting to rename `accelerator` to `parallelizer`. What I meant to say was that comparing to specifying `ddp` through `accelerator="ddp"`, something like `parallelizer="ddp"` makes more sense. Thus I totally agree with this proposal that `accelerator` should be used for specifying the hardware accelerator types (will update my comment to make it more clear).

> Initially, we had considered distributed_type/distributed_strategy naming for the argument. But it wouldn't make sense if SingleDevice plugins are passed.

IMHO distributed training algorithms like `dpp`, `dpp_fully_sharded`, and potential algorithms for recsys and MoE are **NOT** strategies of accelerators. Today, `TrainingTypePlugin`-s are organized internally as strategies of `Accelerator`-s because `Accelerator` delegates most calls to `TrainingTypePlugin`. However, I don't think this delegation makes sense (I vaguely remember there's plan  to change that @ananthsub).

I'd even argue that `Accelerator`-s should be strategies of `TrainingTypePlugin`-s (instead of the other way around), because most distributed training algorithms apply to different hardware accelerators. From Lightning's perspective, maybe `Accelerator` should just be abstracting away data movement and accelerator specific communicators (accelerator specific kernels are handled transparently by PyTorch).

Thus I don't think `accelerator_strategy` is an ideal name, because distributed training algorithms should be strategies of accelerators.

**Comment by ananthsub** [comment]
> IMHO distributed training algorithms like dpp, dpp_fully_sharded, and potential algorithms for recsys and MoE are NOT strategies of accelerators. Today, TrainingTypePlugin-s are organized internally as strategies of Accelerator-s because Accelerator delegates most calls to TrainingTypePlugin. However, I don't think this delegation makes sense (I vaguely remember there's plan to change that @ananthsub).

> I'd even argue that Accelerator-s should be strategies of TrainingTypePlugin-s (instead of the other way around), because most distributed training algorithms apply to different hardware accelerators. From Lightning's perspective, maybe Accelerator should just be abstracting away data movement and accelerator specific communicators (accelerator specific kernels are handled transparently by PyTorch).

Yes, this is the idea being proposed here: https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#heading=h.vv5zw27fkkpe 

Where we flip the accelerator & training type, and make training type (or whatever the new name is) own:
- The hardware accelerator for data movement & device-specific profiling information (e.g. https://github.com/PyTorchLightning/pytorch-lightning/issues/9032)
- Collectives, which can be influenced by the hardware used: https://github.com/PyTorchLightning/pytorch-lightning/issues/7534
- The checkpointing agent (https://github.com/PyTorchLightning/pytorch-lightning/pull/8743)
- The rank information (we could elevate `ClusterEnvironment` as an abstraction, or some reduced interface of this, to be a top-level property of the base training type plugin)
- The optimizer & LR scheduler initialization (and potential rewrapping for things like PostLocalSGD, ZeRO v2, Horovod, or others)
- The precision (since this is highly hardware & comms dependent)

**Comment by ananthsub** [comment]
> Thus I don't think accelerator_strategy is an ideal name, because distributed training algorithms should not be strategies of accelerators.

Ah @yifuwang I misunderstood, my apologies. I think including `parallel` is a great call out as I assume many people toggling this setting would be familiar with data/model/pipeline parallel terminologies. 

@kaushikb11 what do you think of one of `parallel_type`, `parallelizer`, `parallel_strategy` ? this would also reinforces the core assumption of sync-sgd

**Comment by justusschock** [comment]
Imo we should not use the term distributed, since this is not all distributed training (thinking of DP and single device stuff). To me a term like `accelerator_strategy` or `device_strategy` makes more sense.

@ananthsub regarding the deprecation of arguments like `gpus`, `cpus`, `tpu_cores` and `ipus`: I think we have to stick with them for some time. The best we could do is to revisit the docs to not mention them anymore and instead recommend the use of `num_devices` or something similar. But removing them is such a hard change, that we probably cannot do this without having a new major version (e.g. PL 2.x.x).

**Comment by tchaton** [comment]
Personally, I don't like strategy as it doesn't inform the users this is meant to be extended for their own use-case.

Would `device_manager_plugin` makes more sense ?

Best,
T.C

**Comment by yifuwang** [comment]
@tchaton `distributed_backend` sounds good to me too.

> To me a term like accelerator_strategy or device_strategy makes more sense.

@justusschock with the encapsulation of collectives, distributed algorithms in Lightning are close to being device agnostic. IMO the algorithms are not strategies of device usage (as @ananthsub mentioned above, there's plan to eliminate the `Accelerator` -> `TrainingTypePlugin` delegation).

**Comment by justusschock** [comment]
@yifuwang Yes that's true. That's why it's only the strategy on how to use the given devices (the device/hardware handling is implemented by the accelerator).

Just the term `device_manager` is a bit confusing for me, since there it actually indicates the plugin to take over the accelerators responsibilities (even though the `Accelerator` -> `TrainingTypePlugin` delegation might be removed, there still has to be a proper split of responsibilities)

But please let's not use `distributed` here. We had this in the past and it has been proven not to cover everything Lightning offers. Especially single device and DP fall out of this. Let's use something more general here to avoid that we have to change it again...

**Comment by tchaton** [comment]
Makes sense ! 

What about `accelerator_orchestrator` ?

**Comment by Tshimanga** [comment]
İ think maybe we can draw inspiration from similar ideas in the Hadoop/Spark ecosystem. İt sounds like we have two different concerns here: responsibility of interfacing with hardware (CPUs, GPUs, TPUs,...) and a separate Hadoop YARN-like concern of orchestration/resource management. İ would suggest the former to be called a HardwareConnector and the latter a ResourceManager or HardwareOrchestrator?

**Comment by carmocca** [comment]
What's left to do here?
